src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -1246,8 +1257,9 @@ def forward(
             loss = self.loss_function(logits, labels, vocab_size=self.config.vocab_size)
 
         if self.config._attn_implementation == "flash_attention_2":
-            with torch.no_grad():
+            with nullcontext() if self.training or labels is None else torch.no_grad():
```

Comment:
@warner-benjamin I just updated the `config.repad_logits_with_grad` with `self.training`, but I worry that this conditional doesn't make sense. Could you look it over? 
I don't see when we might need grads when not training, so perhaps this can just be:
```suggestion
            with nullcontext() if self.training else torch.no_grad():
```

Reply:
I've reversed this to avoid unnecessarily increasing memory when training in some situations.